components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go

```diff
@@ -0,0 +1,189 @@
+
+	fakecfg.NtripURL = "http://fakeurl"
+	fakecfg.I2CBus = ""
+	_, err = fakecfg.Validate(path)
+	test.That(
+		t,
+		err,
+		test.ShouldBeError,
+		utils.NewConfigValidationFieldRequiredError(path, "i2c_bus"),
+	)
+	fakecfg.I2CBus = testBusName
+	fakecfg.I2CAddr = 0
+	_, err = fakecfg.Validate("path")
+	test.That(
+		t,
+		err,
+		test.ShouldBeError,
+		utils.NewConfigValidationFieldRequiredError(path, "i2c_addr"))
+}
+
+func TestConnect(t *testing.T) {
+	logger := golog.NewTestLogger(t)
+	ctx := context.Background()
+	cancelCtx, cancelFunc := context.WithCancel(ctx)
+	g := rtkI2C{
+		cancelCtx:  cancelCtx,
+		cancelFunc: cancelFunc,
+		logger:     logger,
+	}
+
+	url := "http://fakeurl"
+	username := "user"
+	password := "pwd"
+
+	// create new ntrip client and connect
+	err := g.connect("invalidurl", username, password, 10)
+	g.ntripClient = makeMockNtripClient()
+
+	test.That(t, err, test.ShouldNotBeNil)
+
+	err = g.connect(url, username, password, 10)
+	test.That(t, err, test.ShouldBeNil)
+
+	err = g.getStream("", 10)
+	test.That(t, err, test.ShouldNotBeNil)
+}
+
+func TestPostion(t *testing.T) {
+	logger := golog.NewTestLogger(t)
+	ctx := context.Background()
+	cancelCtx, cancelFunc := context.WithCancel(ctx)
+	g := rtkI2C{
+		cancelCtx:  cancelCtx,
+		cancelFunc: cancelFunc,
+		logger:     logger,
+	}
+
+	mockSensor := &CustomMovementSensor{
+		MovementSensor: &fake.MovementSensor{},
+	}
+
+	mockSensor.PositionFunc = func(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+		return geo.NewPoint(40.7, -73.98), 50.5, nil
+	}
+
+	g.nmeamovementsensor = mockSensor
+
+	// Normal position
+	loc1, alt1, err := g.Position(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, loc1, test.ShouldResemble, geo.NewPoint(40.7, -73.98))
+	test.That(t, alt1, test.ShouldEqual, alt)
+
+	speed1, err := g.LinearVelocity(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, speed1.Y, test.ShouldEqual, speed)
+
+	// Zero position with latitude 0 and longitude 0.
+	mockSensor.PositionFunc = func(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+		return geo.NewPoint(0, 0), 0, nil
+	}
+
+	loc2, alt2, err := g.Position(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, loc2, test.ShouldResemble, geo.NewPoint(0, 0))
+	test.That(t, alt2, test.ShouldEqual, 0)
+
+	speed2, err := g.LinearVelocity(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, speed2.Y, test.ShouldEqual, speed)
+
+	// Position with NaN values.
+	mockSensor.PositionFunc = func(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+		return geo.NewPoint(math.NaN(), math.NaN()), math.NaN(), nil
+	}
+
+	loc3, alt3, err := g.Position(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, math.IsNaN(loc3.Lat()), test.ShouldBeTrue)
```

Comment:
isn't the rtk supposed to return last known position if the position is 0 or NaN?

Reply:
Great question, it is it handled in nmea. And (0,0) is a valid position. NaN is also valid given that rtk received it from nmea meaning something didn't workout and instead of returning (0,0) nmea returned NaN()

Reply:
it's the parser that returns NaN()

Reply:
ohh okay so are we returning that as a valid position in the NMEA postion function? I feel like there should be a check to return the last position if the parser returns NaN